core/shared/src/main/scala-3/org/log4s/LoggerMacros.scala

```diff
@@ -0,0 +1,36 @@
+package org.log4s
+
+import org.slf4j.{ Logger => JLogger }
+import scala.annotation.tailrec
+import scala.language.experimental.macros
+import scala.quoted._
+
+/** Macros that support the logging system.
+  */
+private[log4s] object LoggerMacros {
+
+  def traceTM(logger: Expr[JLogger])(t: Expr[Throwable])(msg: Expr[String])(using qctx: QuoteContext) =
+    '{ if ($logger.isTraceEnabled) $logger.trace($msg, $t) }
```

Comment:
Not as clever as Scala 2's reflective version.  Also doesn't detect constants yet to elide the level check.  But it gets things rolling.